lightly/api/api_workflow_upload_metadata.py

```diff
@@ -168,48 +142,62 @@ def upload_custom_metadata(self,
 
         self.verify_custom_metadata_format(custom_metadata)
 
-        # create a mapping from sample filenames to custom metadata
+
+
+        # For each metadata, we need the corresponding sample_id
+        # on the server. The mapping is:
+        # metadata -> image_id -> filename -> sample_id
+
+        image_id_to_filename = {
+            image_info[COCO_ANNOTATION_KEYS.images_id]:
+                image_info[COCO_ANNOTATION_KEYS.images_filename]
+            for image_info in custom_metadata[COCO_ANNOTATION_KEYS.images]
+        }
+
         samples = self._samples_api.get_samples_by_dataset_id(self.dataset_id)
-        filename_to_metadata = self.index_custom_metadata_by_filename(
-            [sample.file_name for sample in samples],
-            custom_metadata,
-        )
-        if len(filename_to_metadata) != len(custom_metadata[COCO_ANNOTATION_KEYS.images]):
-            raise ValueError(
-                f'There is a mismatch between the number of images '
-                f'in the metadata file ({len(filename_to_metadata)}) and on the '
-                f'server ({len(custom_metadata[COCO_ANNOTATION_KEYS.images])}).'
-            )
+        filename_to_sample_id = {
+            sample.file_name: sample.id
+            for sample in samples
+        }
+
+        upload_requests = []
+        for metadata in custom_metadata[COCO_ANNOTATION_KEYS.custom_metadata]:
+            image_id = metadata[COCO_ANNOTATION_KEYS.custom_metadata_image_id]
+            filename = image_id_to_filename.get(image_id, None)
+            if filename is None:
+                raise ValueError(f"Your custom_metadata file is malformatted. "
+                                 f"There is an annotation with an image_id of "
+                                 f"{{{image_id}}},"
```

Comment:
Why are there three `{{{`?

Reply:
I think we can write shorter and much more precise error messages:
```python
InvalidCustomMetadataError(
    f'No image found for annotation with image_id {image_id}`
)
```